plugin/toerror/toerror.go

```diff
@@ -0,0 +1,127 @@
+//  Copyright 2019 Ingun Jon
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package toerror contains the implementation of the toerror plugin, which generates the deriveToError function.
+//
+// The deriveToError function transforms return type of a function from (a, bool) into (a, error).
+//   deriveToError(e error, f func(...) (T, bool)) func(...) (T, error)
+package toerror
+
+import (
+	"fmt"
+	"go/types"
+	"strings"
+
+	"github.com/awalterschulze/goderive/derive"
+)
+
+// NewPlugin creates a new curry plugin.
+// This function returns the plugin name, default prefix and a constructor for the curry code generator.
+func NewPlugin() derive.Plugin {
+	return derive.NewPlugin("toerror", "deriveToError", New)
+}
+
+// New is a constructor for the curry code generator.
+// This generator should be reconstructed for each package.
+func New(typesMap derive.TypesMap, p derive.Printer, deps map[string]derive.Dependency) derive.Generator {
+	return &gen{
+		TypesMap: typesMap,
+		printer:  p,
+	}
+}
+
+type gen struct {
+	derive.TypesMap
+	printer derive.Printer
+}
+
+func (g *gen) Add(name string, typs []types.Type) (string, error) {
+	if len(typs) != 2 {
+		return "", fmt.Errorf("%s does not have two argument. Expected (error, func (a) (b, bool))", name)
+	}
+	errTyp := typs[0]
+	if !derive.IsError(errTyp) {
+		return "", fmt.Errorf("First parameter should be of type error")
+	}
+	funcTyp := typs[1]
+	sig, ok := funcTyp.(*types.Signature)
+	if !ok {
+		return "", fmt.Errorf("%s, 2nd param %s, is not of type function", name, g.TypeString(typs[1]))
+	}
+	results := sig.Results()
+	if results.Len() <= 0 {
+		return "", fmt.Errorf("%s, given function must return at least 1 bool type", name)
+	}
+	if !types.Identical(results.At(results.Len()-1).Type(), types.Typ[types.Bool]) {
+		return "", fmt.Errorf("%s, given function must return bool as last return type. (got %v)", name, results.String())
+	}
+	return g.SetFuncName(name, errTyp, funcTyp)
```

Comment:
I think we can just say
```
g.SetFuncName(name, funcTyp)
```
The error type is always going to be error, I don't expect duplicates here.

Reply:
I'll see it after 24th
I have a test😫

Reply:
Oh yes please, tests before open source.
I can wait.

Reply:
I believe you are already aware of absence of basic type of `error` in types package.
That was the reason I carried the type all the way from parameter of `Add` to `Generate` even though I knew its just `error`
I resolved it by implementing native `types.Type` acting like basic type `error`
I think its little better this way
what do you think?

Reply:
Very clever, I like it